### ISSUE TITLE
RFC-1008: Formalizing Linkage Strategies in Experiments

### ISSUE THREAD

**Issue by mikkokotila** [Vaquum/Limen#342]
### Objective

Various linkage strategies are widely used in modelling across literature, competitive arenas, and industry. Many modellers automatically fallback to these strategies, yet as it stands, it would be cumbersome, and in some cases not really possible, to achieve these in Limen. So here, we want to formalize the language around linkage strategies, as well as implement them into Limen.

### Summary

Whereas #341 is about single-decoder operations, Linkage Strategies presented here are about multi-decoder operations where one decoder passes something onto the next.

Strategy | The "Link" (What passes from Sensor A → Sensor B?)
-- | --
Residual Linking | The Error. Sensor A passes what it got wrong (residuals) to Sensor B, so B can focus on fixing it.
Weight Linking | The Importance. Sensor A passes a "difficulty score" for each row to Sensor B, telling B which rows to pay more attention to.
Feature Linking | The Prediction. Sensor A passes its answer to Sensor B as a "hint" (a new input column).
Row Linking | The Data. Sensor A passes new "synthetic" training rows (from its own high-confidence predictions) to Sensor B.
Parameter Linking | The Brain. Sensor A passes its learned internal weights to Sensor B so B doesn't have to start learning from scratch.

### Is This Really Necessary?

  Linkage strategies become necessary when single decoder algorithms are insufficient. If
  a single model cannot capture meaningful generalizable patterns, or consistently shows
  poor performance across parameter sweeps, ensembles provide the next layer of
  sophistication. This is not about blindly following industry practice—it's about having
  the capability when single-model approaches hit their limits.

  When does this happen? Industry evidence shows specific scenarios where ensembles
  deliver measurable gains:

  __Financial Trading:__ When market complexity exceeds single-model capacity,
  stacking/blending achieves 90-100% accuracy vs 53-97% for single boosted models. Crypto
  trading reached 81.80% accuracy and 88.43% AUC-ROC using ensemble stacking. Adaptive
  ensembles handle regime changes that single models miss.

  __Kaggle Competitions:__ When marginal gains matter, multi-level stacking is standard in top
   3 winners (3-level stacking in April 2025 Playground winner). 2-5% accuracy
  improvements typical, with top solutions using 5-100+ models to squeeze out performance.

  Why build now? We don't know yet if single models will suffice for our trading
  strategies. But if we wait until we hit single-model limits, we'll need to retrofit
  ensemble support into an architecture not designed for it. We need to design and build our system so that it can naturally accommodate linkage strategies when we decide to implement them.

  __Supporting evidence:__
  - https://link.springer.com/chapter/10.1007/978-3-031-84782-0_14 - 90-100% accuracy
  study
  - https://rpc.cfainstitute.org/research/foundation/2025/chapter-4-ensemble-learning-inve
  stment - Financial markets practice
  - https://developer.nvidia.com/blog/the-kaggle-grandmasters-playbook-7-battle-tested-mod
  eling-techniques-for-tabular-data/ - Competition strategies
  - https://research.google/blog/model-ensembles-are-faster-than-you-think/ - Production
  cost reduction
  - https://arxiv.org/abs/2406.14424 - 2-10x efficiency gains

  ---

### Named Concepts

**Linkage Strategy:** A formal method for connecting two or more sensors where the output or state of an Upstream Sensor modifies the training environment of a Downstream Sensor.

**Link Payload:** The specific information transferred across the link (e.g., a residual vector, a column of predictions, a sample weight vector, or a model checkpoint).

**Upstream/Downstream:** The directional relationship between linked sensors; the Downstream sensor cannot begin training until the Upstream sensor has finalized its payload.

**Residual Chain:** A sequence of sensors connected via Residual Linking, effectively forming a Gradient Boosting Machine where each subsequent sensor reduces the aggregate bias.

**Signal Injection:** The mechanism in Feature Linking where the prediction of a parent sensor is treated as a high-value input feature ("hint") for the child sensor.

**Hard Mining:** The mechanism in Weight Linking where the Downstream sensor is forced to prioritize data points that the Upstream sensor failed to predict correctly.

**Pseudo-Labeling:** The mechanism in Row Linking where high-confidence predictions on unlabelled/test data are converted into synthetic "ground truth" training rows for the next sensor.

### Outline

There is a wide range of strategies that can improve sensor performance. The only way to truly find out what works and what doesn't, is to expose it so that it can be included in experiments and cohorts, and eventually "tested" downstream in live or simulated trading.

There are three key considerations here:

- The way in which the user is able to introduce and control linkages
- The way in which the linkages are presented in the yml manifest
- The way in which the linkages are captured in the logs

### Out of Scope

Some of the linkages proposed here might be too heavy of a lift to be included in the initial build. But it's useful to know that these are the things that we have to eventually build in, so that we don't now make sidesteps that make implementation untenable later. It might also end up that everything presented here will end up being included in the initial build scope. 

Based on feasibility analysis:

__Parameter Linking__ is explicitly out of scope for this RFC. Research across 40+
  sources covering Kaggle competitions, financial trading, and production ML systems shows
   **zero usage of parameter linking for tabular data**. It is primarily used in
  NLP/vision domains with foundation models (BERT, ImageNet) where pre-trained weights
  transfer well across tasks. For tabular data, features are dataset-specific and models
  train quickly from scratch—making parameter linking unnecessary. The four remaining
  linkage strategies (Residual, Weight, Feature, Row) have strong industry evidence and
  clear value.

### User Experience

The standard Limen user-experience should not change because of these. These things should simply fall into that pattern cohesively and comprehensively.

### New Data

Minimal extensions to existing experiment log schema. All changes are additive and backward compatible with existing design of `reference_architecture`.

#### Experiment Log Extensions

**1. Ensemble Composition Metadata**
- Linkage type identifier (residual/weight/feature/row)
- Component model identifiers and dependency order (upstream → downstream)
- Stored in experiment metadata section of log

**2. Nested Parameters (Dot Notation)**
- Flat keys with dots: `level0_0.learning_rate`, `level0_1.C`, `level1.alpha`
- No schema changes to ParamSpace or manifest_core
- Requires simple YAML flattening utility if YAML input has nested structure
- Example:
  ```python
  # Stored in experiment log as flat dict
  {
    'level0_0.learning_rate': 0.01,
    'level0_0.max_depth': 3,
    'level0_1.C': 1.0,
    'level1.alpha': 0.1
  }
  ```

**3. Link-Specific Configuration**
- **Weight Linking:** Per-model weights array (`weights: [1.0, 1.5, 0.8]`)
- **Feature Linking:** Meta-model configuration (stacking/blending settings)
- **Row Linking:** Confidence thresholds, sampling ratios, bootstrap settings

**4. Per-Component Random Seeds**
- Explicit seeds for each model in ensemble
- Ensures deterministic reconstruction of all components
- Extends existing seed management from RFC-1012

**5. Metrics Structure**
- **Component-level metrics:** Individual model performance (for debugging/analysis)
- **Ensemble aggregate metrics:** Combined performance (primary optimization target)
- **Linkage-specific metrics:** Cascade early-exit rates, pseudo-labeling confidence distributions
- Example:
  ```python
  {
    'aggregate_accuracy': 0.92,
    'component_0_accuracy': 0.88,
    'component_1_accuracy': 0.90,
    'ensemble_type': 'soft_voting'
  }
  ```
---
### Support

Falls under regular support.

### Market Desirability

Nothing is controversial here. Everything falls under standard modelling practices, perhaps here more so even than with #1007. These practices are standard in Kaggle competitions, industry application, and other competitive arenas. Also, very importantly, we would not want ourselves to be live trading without these basic linkage primitives.

### Risks

The primary risk is implementing ensemble capabilities before validating that single models are insufficient for our trading strategies. If single models perform adequately, the ensemble infrastructure (manifest_core modification, ensemble classes) may be underutilized. Mitigation: The ~10 line manifest_core change is small and backward compatible; ensemble classes can be added incrementally as needs emerge.

### Drawbacks

**1. Computational Cost Multiplication:**  Training 3-5 models per permutation means 3-5x computation time and memory. Stacking with out-of-fold predictions is even more expensive. Inherent tradeoff with no simple mitigation.

**2. Hyperparameter Space Explosion:**  Nested parameters multiply search space exponentially. Requires aggressive pruning strategies (RFC-1010) to remain tractable.

**3. Overfitting Risk:**  Multi-level ensembles (especially stacking) can overfit validation data if meta-model memorizes patterns rather than learns generalizable rules. Requires validation discipline and proper cross-validation.

**4. Deterministic Reconstruction Complexity:**  Ensembles require: explicit seeds for all components, deterministic training order, reproducible OOF splitting, exact OOF prediction matching. Single missed seed breaks RFC-1012 reconstruction guarantees.

---

### Alternatives

There are no real alternative here as the single decoder based experiments are already implemented. Besides any external/manual ensemble is already in place through #307 

### Unresolved Questions

Are there other linkage practices that we should consider? 

### Success Metrics

A complete implementation, without breaking standard Limen user-experience.

### Business Release Date

TBA

### Design

**This is not a complete design, but an outline ensuring our current system can accommodate linkage strategies without architectural changes.**  Based on feasibility analysis of manifest_core.py, ParamSpace, and UniversalExperimentLoop, we confirm that linkage strategies can be implemented through extension, not modification.

---

#### Design Overview

**Single Implementation Approach: Ensemble Reference Architecture Classes**

All linkage strategies are implemented as new reference architecture classes using **dot notation for parameter namespacing**. The key insight: `model_0.learning_rate` is treated as a regular parameter name by ParamSpace and UEL—no schema changes needed.

**Use cases:** Voting, Stacking, Residual, Bagging, Cascading, Pseudo-labeling

**Why this works:**
- ParamSpace treats `'model_0.learning_rate'` as just another key
- UEL passes flat `round_params` dict unchanged
- Ensemble classes extract nested params via string parsing (`k.startswith('model_0.')`)
- Small modification to manifest_core.py to support `**kwargs` pattern 

---

##### Example 1: Voting Ensemble

**YAML Configuration:**
```yaml
manifest:
  reference_architecture: limen.sfd.reference_architecture.VotingEnsemble

params:
  # Model-specific params with dot notation
  model_0.learning_rate: [0.01, 0.1]
  model_0.max_depth: [2, 3]
  model_1.C: [1.0, 10.0]
  model_1.max_iter: [100, 200]

  # Ensemble configuration
  model_types: [['XGBoostRegressor', 'LogRegBinary']]
  weights: [[1.0, 1.5]]
  ensemble_method: ['soft_voting']
```

**Reference Architecture Implementation:**
```python
def voting_ensemble(data, **all_params):
    # Extract model-specific params via string parsing
    model_0_params = {k.split('.', 1)[1]: v for k, v in all_params.items()
                      if k.startswith('model_0.')}
    model_1_params = {k.split('.', 1)[1]: v for k, v in all_params.items()
                      if k.startswith('model_1.')}

    # Get ensemble config
    model_types = all_params['model_types']
    weights = all_params['weights']

    # Train models
    model_0 = XGBoostRegressor(data, **model_0_params)
    model_1 = LogRegBinary(data, **model_1_params)

    # Aggregate predictions
    final_pred = (weights[0] * model_0.predict(data) +
                  weights[1] * model_1.predict(data)) / sum(weights)

    return {'aggregate_accuracy': ..., 'component_0_acc': ..., 'component_1_acc': ...}
```

**How it flows:**
1. ParamSpace samples: `{'model_0.learning_rate': 0.01, 'model_0.max_depth': 2, 'model_1.C': 1.0, ...}`
2. UEL passes flat dict to manifest.run_model()
3. manifest.run_model() sees `**kwargs` in voting_ensemble signature → passes all params
4. voting_ensemble extracts nested params and trains models

---

##### Example 2: Stacking Ensemble
```yaml
manifest:
  reference_architecture: limen.sfd.reference_architecture.StackingEnsemble

params:
  # Nested params using dot notation (flat storage)
  level0_0.learning_rate: [0.01, 0.1]
  level0_0.max_depth: [2, 3, 4]
  level0_1.C: [1.0, 10.0]
  level0_1.penalty: ['l1', 'l2']
  level1.alpha: [0.1, 1.0]

  # Ensemble config
  level0_models: [['XGBoostRegressor', 'LogRegBinary']]  # Fixed list
  level1_model: ['Ridge']                                 # Fixed
  cv_folds: [5]
```

**How it works:**
- ParamSpace samples flat dict: `{'level0_0.learning_rate': 0.01, 'level0_1.C': 1.0, ...}`
- StackingEnsemble extracts nested params via string splitting
- Trains level-0 models with out-of-fold predictions
- Trains level-1 meta-model on OOF predictions
- Returns aggregate metrics + component metrics

**StackingEnsemble reference architecture (sketch):**
```python
def stacking_ensemble(data, **all_params):
    # Extract params for each component
    level0_0 = {k.split('.', 1)[1]: v for k, v in all_params.items()
                if k.startswith('level0_0.')}
    level0_1 = {k.split('.', 1)[1]: v for k, v in all_params.items()
                if k.startswith('level0_1.')}
    level1 = {k.split('.', 1)[1]: v for k, v in all_params.items()
              if k.startswith('level1.')}

    # Train level-0 models with OOF predictions
    oof_preds_0 = train_with_cv(XGBoostRegressor, data, **level0_0)
    oof_preds_1 = train_with_cv(LogRegBinary, data, **level0_1)

    # Train level-1 meta-model on OOF predictions
    meta_features = np.column_stack([oof_preds_0, oof_preds_1])
    meta_model = Ridge(**level1).fit(meta_features, data['y_train'])

    # Evaluate and return metrics
    return {'aggregate_accuracy': ..., 'component_0_acc': ..., ...}
```

---

#### Required Changes

**1. Small Modification to manifest_core.py**

Current `run_model()` uses signature inspection for models with explicit parameters. Ensemble classes need `**kwargs` pattern to accept dot-notation parameters.

**Add `**kwargs` support:**
```python
def run_model(self, data: dict, round_params: dict[str, Any]) -> dict:
    if self.model_function is None:
        raise ValueError('Model function not configured...')

    sig = inspect.signature(self.model_function)

    # Check if function accepts **kwargs
    has_var_keyword = any(p.kind == inspect.Parameter.VAR_KEYWORD
                          for p in sig.parameters.values())

    if has_var_keyword:
        # Ensemble pattern: pass all params
        round_results = self.model_function(data, **round_params)
    else:
        # Existing pattern: signature inspection (backward compatible)
        model_kwargs = {}
        for param_name, param_obj in sig.parameters.items():
            if param_name == 'data':
                continue
            if param_name in round_params:
                model_kwargs[param_name] = round_params[param_name]
            elif param_obj.default != inspect.Parameter.empty:
                model_kwargs[param_name] = param_obj.default
            else:
                raise ValueError(...)

        round_results = self.model_function(data, **model_kwargs)

    return round_results
```

**2. New Ensemble Reference Architecture Classes**
- VotingEnsemble, StackingEnsemble, ResidualEnsemble, BaggingEnsemble, CascadingEnsemble, PseudoLabelingEnsemble
- Each implements standard interface: `def ensemble_name(data, **all_params) -> dict`

**3. Optional: YAML Flattening Utility**
- If users write nested YAML, flatten to dot notation
- Not required if users write flat params with dots directly

**4. Seed Management**
- Explicit seeds for each component model
- Ensures deterministic OOF splitting for stacking

---

#### Key Architectural Validations

**No changes to ParamSpace**
- Dot notation keys (`model_0.learning_rate`) treated as regular keys
- Sampling/Search/Pruning logic unchanged
- Works seamlessly with current implementation

**No changes to UEL core**
- round_params remains flat dict
- Experiment loop unchanged
- Still calls `self.model()` once per permutation

**Small modification to manifest_core.py**
- Add `**kwargs` support to `run_model()`
- Backward compatible (checks parameter kind first)
- Existing models with explicit parameters unchanged

**No manifest schema changes**
- reference_architecture remains single path (not list)
- params remains flat dict structure
- Backward compatible with all existing SFDs

**Trainer reconstruction**
- Treats ensemble as single reference architecture
- Passes all params to ensemble class
- Ensemble handles internal reconstruction

---

#### Linkage Strategy Mapping

| Linkage Type | Implementation |
|--------------|----------------|
| **Residual** | ResidualEnsemble (train model A → calculate residuals → train model B on residuals → combine) |
| **Weight** | VotingEnsemble (weighted voting/averaging across models with configurable weights) |
| **Feature** | StackingEnsemble (OOF predictions as features), CascadingEnsemble (early exit with confidence thresholds) |
| **Row** | BaggingEnsemble (bootstrap sampling), PseudoLabelingEnsemble (high-confidence synthetic labels) |
| **Parameter** | Out of scope - zero industry evidence for tabular ML |

---

#### Architectural Change Assessment

**Changes Required:**
1. **manifest_core.py modification**: Add `**kwargs` support to `run_model()`
   - Localized to single method
   - Backward compatible (checks parameter kind, falls back to existing logic)
   - No changes to core experiment loop, parameter system, or manifest schema

2. **New ensemble classes** (pure extension): VotingEnsemble, StackingEnsemble, etc.
   - Standard reference architecture pattern
   - No changes to existing code

---

#### Next Steps for Full Design

When ready to implement, full design should specify:
1. Detailed specification of manifest_core.run_model() `**kwargs` modification
2. Ensemble class interfaces and standard contracts
3. Experiment log extensions for ensemble metadata (composition, component metrics)
4. Seed management protocol for deterministic OOF splitting
5. Validation strategy to prevent overfitting (especially for stacking)
6. Integration with RFC-1010 pruning strategies for exponential hyperparameter spaces
7. Trainer reconstruction logic for ensemble reference architectures
8. YAML flattening utility specification (if supporting nested input)

---

### Work Phases
TBA

#### Non-coding

- [ ] 

#### Coding

- [ ]
